File: src/apple/Packages/MapleUI/Sources/MapleUI/Organisms/MuiSearchResults.swift
// MuiSearchResults.swift — Maple UI Organisms · Collections
// (unified-component-catalog.md §4.1). Paginated result grid with states,
// built from Collection Grid, Empty State, Progress.

import SwiftUI

public struct MuiSearchResults: View {
    public let items: [MuiCollectionItem]
    public let loading: Bool
    public let query: String
    public let totalCount: Int
    public let page: Int
    public let pageCount: Int
    public let columns: Int
    @Binding public var selectedIds: [String]
    public let opened: ((String) -> Void)?
    public let pageChanged: ((Int) -> Void)?

    public init(
        items: [MuiCollectionItem],
        loading: Bool = false,
        query: String = "",
        totalCount: Int = 0,
        page: Int = 1,
        pageCount: Int = 1,
        columns: Int = 4,
        selectedIds: Binding<[String]> = .constant([]),
        opened: ((String) -> Void)? = nil,
        pageChanged: ((Int) -> Void)? = nil
    ) {
        self.items = items
        self.loading = loading
        self.query = query
        self.totalCount = totalCount
        self.page = page
        self.pageCount = pageCount
        self.columns = columns
        self._selectedIds = selectedIds
        self.opened = opened
        self.pageChanged = pageChanged
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: MuiTokens.spacingSm) {
            header

            if loading {
                MuiProgress(shape: .bar, label: "Searching…")
                    .padding(.horizontal, MuiTokens.spacingMd)
            } else if items.isEmpty {
                MuiEmptyState(
                    icon: "magnifyingglass",
                    title: "No results",
                    message: query.isEmpty ? "Try a different search." : "No matches for \u{201C}\(query)\u{201D}."
                )
            } else {
                MuiCollectionGrid(items: items, columns: columns, selectedIds: $selectedIds, opened: opened)
                pagination
            }
        }
    }

    private var header: some View {
        HStack {
            MuiText(query.isEmpty ? "\(totalCount) results" : "\(totalCount) results for \u{201C}\(query)\u{201D}", variant: .body, color: .muted)
            Spacer()
        }
        .padding(.horizontal, MuiTokens.spacingMd)
    }

    private var pagination: some View {
        HStack {
            MuiButton(label: "Previous", variant: .ghost, size: .sm, disabled: page <= 1) {
                pageChanged?(page - 1)
            }
            Spacer()
            MuiText("Page \(page) of \(pageCount)", variant: .body, color: .muted)
            Spacer()
            MuiButton(label: "Next", variant: .ghost, size: .sm, disabled: page >= pageCount) {
                pageChanged?(page + 1)
            }
        }
        .padding(.horizontal, MuiTokens.spacingMd)
        .padding(.bottom, MuiTokens.spacingSm)
    }
}

#Preview("MuiSearchResults — Populated") {
    struct Demo: View {
        @State private var selected: [String] = []
        var body: some View {
            MuiSearchResults(
                items: (1...8).map { MuiCollectionItem(id: "\($0)", url: nil, alt: "Result \($0)") },
                query: "iceland",
                totalCount: 42,
                page: 2,
                pageCount: 6,
                columns: 4,
                selectedIds: $selected
            )
            .frame(height: 420)
            .background(MuiTokens.bg)
        }
    }
    return Demo()
}

#Preview("MuiSearchResults — Loading / Empty") {
    VStack(spacing: 0) {
        MuiSearchResults(items: [], loading: true, query: "iceland").frame(height: 120)
        MuiDivider()
        MuiSearchResults(items: [], query: "zzzz").frame(height: 160)
    }
    .background(MuiTokens.bg)
}

File: src/windows/Maple.WinUI/MapleUI/Organisms/MuiSearchResults.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Search Results organism (unified-component-catalog.md
    /// §4.1, "Search Results" row: "Paginated result grid with states",
    /// built from Collection Grid, Empty State, Progress) — a
    /// <see cref="MuiCollectionGrid"/> with a result-count line, a
    /// determinate <see cref="MuiProgress"/> while a page loads, and
    /// Prev/Next paging controls. Empty state and the loading spinner for
    /// an EMPTY page are the grid's own (it already composes Empty State
    /// and Spinner) — this organism's own Progress is specifically the
    /// "loading page N of M, results already on screen" state a plain
    /// spinner-on-empty doesn't cover.
    /// </summary>
    public sealed class MuiSearchResults : ContentControl
    {
        public static readonly DependencyProperty ItemsProperty =
            DependencyProperty.Register(nameof(Items), typeof(IReadOnlyList<MuiCollectionGridItem>), typeof(MuiSearchResults),
                new PropertyMetadata(null, (d, _) => ((MuiSearchResults)d).Rebuild()));

        public static readonly DependencyProperty SelectedIdsProperty =
            DependencyProperty.Register(nameof(SelectedIds), typeof(IReadOnlyList<string>), typeof(MuiSearchResults),
                new PropertyMetadata(null, (d, _) => ((MuiSearchResults)d).Rebuild()));

        public static readonly DependencyProperty TotalCountProperty =
            DependencyProperty.Register(nameof(TotalCount), typeof(int), typeof(MuiSearchResults),
                new PropertyMetadata(0, (d, _) => ((MuiSearchResults)d).Rebuild()));

        public static readonly DependencyProperty PageProperty =
            DependencyProperty.Register(nameof(Page), typeof(int), typeof(MuiSearchResults),
                new PropertyMetadata(1, (d, _) => ((MuiSearchResults)d).Rebuild()));

        public static readonly DependencyProperty PageCountProperty =
            DependencyProperty.Register(nameof(PageCount), typeof(int), typeof(MuiSearchResults),
                new PropertyMetadata(1, (d, _) => ((MuiSearchResults)d).Rebuild()));

        public static readonly DependencyProperty IsLoadingPageProperty =
            DependencyProperty.Register(nameof(IsLoadingPage), typeof(bool), typeof(MuiSearchResults),
                new PropertyMetadata(false, (d, _) => ((MuiSearchResults)d).Rebuild()));

        public IReadOnlyList<MuiCollectionGridItem>? Items
        {
            get => (IReadOnlyList<MuiCollectionGridItem>?)GetValue(ItemsProperty);
            set => SetValue(ItemsProperty, value);
        }

        public IReadOnlyList<string>? SelectedIds
        {
            get => (IReadOnlyList<string>?)GetValue(SelectedIdsProperty);
            set => SetValue(SelectedIdsProperty, value);
        }

        public int TotalCount { get => (int)GetValue(TotalCountProperty); set => SetValue(TotalCountProperty, value); }
        public int Page { get => (int)GetValue(PageProperty); set => SetValue(PageProperty, value); }
        public int PageCount { get => (int)GetValue(PageCountProperty); set => SetValue(PageCountProperty, value); }
        public bool IsLoadingPage { get => (bool)GetValue(IsLoadingPageProperty); set => SetValue(IsLoadingPageProperty, value); }

        public event EventHandler<IReadOnlyList<string>>? SelectionChanged;
        public event EventHandler<string>? ItemActivated;
        public event EventHandler<int>? PageRequested;

        private readonly StackPanel _root = new() { Orientation = Orientation.Vertical, Spacing = 12 };
        private readonly MuiText _countText = new() { Variant = MuiTextVariant.Eyebrow, ColorRole = MuiTextColorRole.Muted };
        private readonly MuiProgress _pageProgress = new() { IsIndeterminate = true, Visibility = Visibility.Collapsed };
        private readonly MuiCollectionGrid _grid = new() { EmptyTitle = "No results", EmptyMessage = "Try a different search or filter." };
        private readonly StackPanel _pager = new() { Orientation = Orientation.Horizontal, Spacing = 8, HorizontalAlignment = HorizontalAlignment.Center };
        private readonly MuiButton _prev = new() { Variant = MuiButtonVariant.Secondary, Label = "Previous" };
        private readonly MuiText _pageText = new() { Variant = MuiTextVariant.Body, ColorRole = MuiTextColorRole.Muted };
        private readonly MuiButton _next = new() { Variant = MuiButtonVariant.Secondary, Label = "Next" };

        public MuiSearchResults()
        {
            _root.Children.Add(_countText);
            _root.Children.Add(_pageProgress);
            _root.Children.Add(_grid);
            _pager.Children.Add(_prev);
            _pager.Children.Add(_pageText);
            _pager.Children.Add(_next);
            _root.Children.Add(_pager);
            Content = _root;

            _grid.SelectionChanged += (_, ids) => SelectionChanged?.Invoke(this, ids);
            _grid.ItemActivated += (_, id) => ItemActivated?.Invoke(this, id);
            _prev.Click += (_, _) => PageRequested?.Invoke(this, Page - 1);
            _next.Click += (_, _) => PageRequested?.Invoke(this, Page + 1);

            Rebuild();
        }

        private void Rebuild()
        {
            var items = Items ?? Array.Empty<MuiCollectionGridItem>();
            _countText.Text = TotalCount == 1 ? "1 result" : $"{TotalCount:N0} results";
            _pageProgress.Visibility = IsLoadingPage && items.Count > 0 ? Visibility.Visible : Visibility.Collapsed;

            _grid.Items = items;
            _grid.SelectedIds = SelectedIds;
            _grid.IsLoading = IsLoadingPage && items.Count == 0;

            _pager.Visibility = PageCount > 1 ? Visibility.Visible : Visibility.Collapsed;
            _prev.IsEnabled = Page > 1;
            _next.IsEnabled = Page < PageCount;
            _pageText.Text = $"Page {Page} of {PageCount}";
        }
    }
}
